File: legacy-nodes/src/status-server.ts
import { PeerId } from '@libp2p/interface'
import { Libp2pType } from './types.js'
import { WebSocketServer } from 'ws';
import { isEqual } from './helpers.js';
import { toString } from 'uint8arrays'
import { multiaddr } from '@multiformats/multiaddr'

interface Stream {
  protocol: string,
  direction: string
}

interface Streams {
  [key: string]: Stream[]
}

interface PeerScores {
  [key: string]: number
}

interface RTTs {
  [key: string]: number
}

type TopicsPeers = Record<string, string[]>

export interface Update {
  containerId?: string
  peerId?: string,
  type?: string
  subscribersList?: TopicsPeers,
  pubsubPeers?: string[],
  meshPeersList?: TopicsPeers,
  fanoutList?: TopicsPeers,
  libp2pPeers?: string[],
  connections?: string[],
  protocols?: string[],
  streams?: Streams,
  multiaddrs?: string[],
  topics?: string[],
  dhtPeers?: string[],
  lastMessage?: string,
  peerScores?: PeerScores,
  rtts?: RTTs
  connectTime?: number
}

export class StatusServer {
  // server details
  private server: Libp2pType
  private type: string
  private peerId: PeerId
  private topics: string[]
  private containerId: string = ''
  private perfBytes: number = 0

  // libp2p
  private lastPeerId: string = ''
  private lastType: string = ''
  private lastLibp2pPeers: string[] = []
  private lastConnections: string[] = []
  private lastProtocols: string[] = []
  private lastStreams: Streams = {}
  private lastMultiaddrs: string[] = []
  private lastRTTs: RTTs = {} // round trip times

  // pubsub
  private lastTopics: string[] = []
  private lastSubscribersList: TopicsPeers = {}
  private lastPubsubPeers: string[] = []
  private lastMeshPeersList: TopicsPeers = {}
  private lastFanoutList: TopicsPeers = {}
  private lastMessage: string = ''
  private lastPeerScores: PeerScores = {}
  private _message: string = ''

  // dht
  private lastDhtPeers: string[] = []

  // perf
  private connectTime: number = 0
  private lastConnectTime: number = 0

  private wss: WebSocketServer
  private wssAlive: boolean
  private started: boolean
  private scheduleUpdate!: NodeJS.Timeout
  private updateCount: number = 0
  private updateIntervalMs: number = 100
  private updatesBeforeFullData: number = 100 // (this.updateIntervalMs * 100) // 10s
  private rttUpdateIntervalMs: number = 1000

  constructor(server: Libp2pType, type: string, topics: string[], perfBytes: number) {
    this.server = server
    this.type = type
    this.peerId = server.peerId
    this.topics = topics
    this.perfBytes = perfBytes

    this.wssAlive = false
    this.started = false

    this.wss = new WebSocketServer({ host: '0.0.0.0', port: 80 });
    this.wssSetup(this.wss)

    server.addEventListener('connection:open', this.handleConnectionEvent)
    server.addEventListener('connection:close', this.handleConnectionEvent)
    server.addEventListener('connection:prune', this.handleConnectionEvent)
    server.addEventListener('self:peer:update', this.handleSelfPeerUpdate)

    this.rttSetup()
    this.started = true
  }

  private handleSelfPeerUpdate = async (evt: CustomEvent) => {
    const update = await this.fullUpdate()
    await this.sendUpdate(update)
  }

  private handleConnectionEvent = async (evt: CustomEvent) => {
    const connectionList = this.server.getConnections()
    const connections = connectionList.map((connection) => connection.remotePeer.toString())

    const update: Update = {
      connections
    }
    await this.sendUpdate(update)
  }

  private handlePubsubMessageEvent = async (evt: CustomEvent) => {
    if (!this.topics.includes(evt.detail.topic)) {
      console.log('unknown topic', evt.detail.topic)
      return
    }

    this.message = toString(evt.detail.data)

    const update: Update = {
      lastMessage: this.message
    }

    await this.sendUpdate(update)
  }

  private getPeerScores = (): PeerScores => {
    return {}
  }

  private wssSetup = (wss: WebSocketServer) => {
    const self = this

    wss.on('close', function close() {
      console.log('stopping scheduled updates')
      clearInterval(self.scheduleUpdate);
    });

    wss.on('connection', async function connection(ws) {
      console.log('new websocket connection')
      ws.on('error', console.error);
      ws.on('pong', self.heartbeat);
      ws.on('message', async function incoming(msg) {
        const newMessage = JSON.parse(msg.toString())

        switch (newMessage.mType) {
          case 'set-id': {
            console.log('setting container id', newMessage.message)
            self.containerId = newMessage.message
            const update = await self.fullUpdate()
            await self.sendUpdate(update)
            break
          }

          case 'info': {
            const update: Partial<Update> = {}
            self.lastPeerScores = self.getPeerScores()
            update.peerScores = self.lastPeerScores

            ws.send(JSON.stringify(update))
            break;
          }

          case 'publish': {
            break;
          }

          case 'connect': {
            console.log('connect msg', newMessage)
            try {
              const ma = multiaddr(newMessage.message)
              console.log('dialing', ma)
              const start = process.hrtime();
              await self.server.dial(ma, { signal: AbortSignal.timeout(10_000) })
              if (self.perfBytes) {
                try {
                  for await (const output of self.server.services.perf.measurePerformance(ma, self.perfBytes, self.perfBytes)) {
                    console.log('perf', output)
                  }
                } catch (err: any) {
                  console.error('Error measuring performance:', err)
                }
              }
              const [seconds, nanoseconds] = process.hrtime(start);
              self.connectTime = seconds * 1000 + nanoseconds / 1e6;
              console.log('dialed', ma)
            } catch (e) {
              console.log(e)
            }

            break;
          }

          default:
            console.log('unknown message type', newMessage.type)
            break;
        }
      });

      ws.send(JSON.stringify(await self.deltaUpdate()))
      self.scheduleUpdate = self.scheduleUpdates()
    })

    this.wssAlive = true
  }

  public get isStarted(): boolean {
    return this.started
  }

  get message(): string {
    return this._message
  }

  set message(message: string) {
    this._message = message
  }

  private getAllSubscribersAllTopics = (): TopicsPeers => {
    return {}
  }

  private getAllMeshPeersAllTopics = (): TopicsPeers => {
    return {}
  }

  private getFanoutPeers = (): TopicsPeers => {
    return {}
  }

  private deltaUpdate = async (): Promise<Update> => {
    const update: Update = {}

    // libp2p
    const peerId = this.peerId.toString()
    if (peerId !== this.lastPeerId) {
      this.lastPeerId = peerId
      update.peerId = peerId
    }

    const type = this.type
    if (type !== this.lastType) {
      this.lastType = type
      update.type = type

    }

    // const topics = this.topics
    // if (topics !== this.lastTopics) {
    //   this.lastTopics = topics
    //   update.topics = topics
    // }

    const libp2pPeerList = await this.server.peerStore.all()
    const libp2pPeers = libp2pPeerList.map((peer) => peer.id.toString())
    if (!isEqual(libp2pPeers, this.lastLibp2pPeers)) {
      this.lastLibp2pPeers = libp2pPeers
      update.libp2pPeers = libp2pPeers
    }

    // connections have own handler
    // const connectionList = this.server.getConnections()
    // const connections = connectionList.map((connection) => connection.remotePeer.toString())
    // if (!isEqual(connections, this.lastConnections)) {
    //   this.lastConnections = connections
    //   update.connections = connections
    // }

    const streams = this.getStreams()
    if (!isEqual(streams, this.lastStreams)) {
      this.lastStreams = streams
      update.streams = streams
    }

    const protocols = this.server.getProtocols()
    if (!isEqual(protocols, this.lastProtocols)) {
      this.lastProtocols = protocols
      update.protocols = protocols
    }

    const multiaddrList = this.server.getMultiaddrs()
    const multiaddrs = multiaddrList.map((ma) => ma.toString())
    if (!isEqual(multiaddrs, this.lastMultiaddrs)) {
      this.lastMultiaddrs = multiaddrs
      update.multiaddrs = multiaddrs
    }

    // pubsub
    // pubsub topics
    const topics: string[] = []
    if (!isEqual(topics, this.lastTopics)) {
      this.lastTopics = topics
      update.topics = topics
    }

    // pubsub topic subscribers
    const subscribersList = this.getAllSubscribersAllTopics()
    if (!isEqual(subscribersList, this.lastSubscribersList)) {
      this.lastSubscribersList = subscribersList
      update.subscribersList = subscribersList
    }

    // pubsub peers
    const pubsubPeers: string[] = []
    if (!isEqual(pubsubPeers, this.lastPubsubPeers)) {
      this.lastPubsubPeers = pubsubPeers
      update.pubsubPeers = pubsubPeers
    }

    // pubsub mesh peers
    const meshPeersList = this.getAllMeshPeersAllTopics()
    if (!isEqual(meshPeersList, this.lastMeshPeersList)) {
      this.lastMeshPeersList = meshPeersList
      update.meshPeersList = meshPeersList
    }

    // gossipsub fanout peers
    const fanoutList = this.getFanoutPeers()
    if (!isEqual(fanoutList, this.lastFanoutList)) {
      this.lastFanoutList = fanoutList
      update.fanoutList = fanoutList
    }

    // gossipsub peer scores
    const peerScores = this.getPeerScores()
    if (!isEqual(peerScores, this.lastPeerScores)) {
      this.lastPeerScores = peerScores
      update.peerScores = peerScores
    }

    // Round trip times
    const rtts = this.getRTTs()
    if (!isEqual(rtts, this.lastRTTs)) {
      this.lastRTTs = rtts
      update.rtts = rtts
    }

    // messages have own handler
    // if (this.lastMessage !== this._message) {
    //   this.lastMessage = this._message
    //   update.lastMessage = this.lastMessage
    // }

    // dht
    // @ts-ignore-next-line
    const dhtPeers: string[] = []
    if (!isEqual(dhtPeers, this.lastDhtPeers)) {
      this.lastDhtPeers = dhtPeers
      update.dhtPeers = dhtPeers
    }

    if (!isEqual(this.connectTime, this.lastConnectTime)) {
      update.connectTime = this.connectTime
    }
    return update
  }

  private fullUpdate = async (): Promise<Update> => {
    const update: Update = {}

    // libp2p
    const peerId = this.peerId.toString()
    this.lastPeerId = peerId
    update.peerId = peerId

    const type = this.type
    this.lastType = type
    update.type = type

    // const topic = this.topic
    // this.lastTopic = topic
    // update.topic = topic

    const libp2pPeerList = await this.server.peerStore.all()
    const libp2pPeers = libp2pPeerList.map((peer) => peer.id.toString())
    this.lastLibp2pPeers = libp2pPeers
    update.libp2pPeers = libp2pPeers

    const connectionList = this.server.getConnections()
    const connections = connectionList.map((connection) => connection.remotePeer.toString())
    this.lastConnections = connections
    update.connections = connections

    const streams = this.getStreams()
    this.lastStreams = streams
    update.streams = streams

    const protocols = this.server.getProtocols()
    this.lastProtocols = protocols
    update.protocols = protocols

    const multiaddrList = this.server.getMultiaddrs()
    const multiaddrs = multiaddrList.map((ma) => ma.toString())
    this.lastMultiaddrs = multiaddrs
    update.multiaddrs = multiaddrs

    // pubsub
    const topics: string[] = []
    this.lastTopics = topics
    update.topics = topics

    const subscribersList = this.getAllSubscribersAllTopics()
    this.lastSubscribersList = subscribersList
    update.subscribersList = subscribersList

    const pubsubPeers: string[] = []
    this.lastPubsubPeers = pubsubPeers
    update.pubsubPeers = pubsubPeers

    const meshPeersList = this.getAllMeshPeersAllTopics()
    this.lastMeshPeersList = meshPeersList
    update.meshPeersList = meshPeersList

    const fanoutList = this.getFanoutPeers()
    this.lastFanoutList = fanoutList
    update.fanoutList = fanoutList

    this.lastPeerScores = this.getPeerScores()
    update.peerScores = this.lastPeerScores

    this.lastRTTs = this.getRTTs()
    update.rtts = this.lastRTTs

    this.lastMessage = this._message
    update.lastMessage = this.lastMessage

    // dht
    const dhtPeers: string[] = []
    this.lastDhtPeers = dhtPeers
    update.dhtPeers = dhtPeers

    // perf connect
    this.lastConnectTime = this.connectTime
    update.connectTime = this.lastConnectTime

    return update
  }

  private sendUpdate = async (update: Update) => {
    if (update.type || update.subscribersList || update.pubsubPeers || update.meshPeersList || update.libp2pPeers || update.connections || update.protocols || update.streams || update.multiaddrs || update.dhtPeers || update.lastMessage || update.peerScores || update.rtts || update.connectTime) {
      update.containerId = this.containerId

      this.wss.clients.forEach((ws) => {
        if (ws.readyState === ws.OPEN) {
          ws.send(JSON.stringify(update));
        } else {
          console.log('websocket not open')
        }
      })
    }
  }

  private scheduleUpdates = () => {
    const updateInterval = setInterval(async () => {
      this.updateCount++

      if (this.updateCount >= this.updatesBeforeFullData) {
        this.updateCount = 0
        const update = await this.fullUpdate()
        // console.log('full update', update)
        await this.sendUpdate(update)
      } else {
        const update = await this.deltaUpdate()
        await this.sendUpdate(update)
      }
    }, this.updateIntervalMs)

    return updateInterval
  }

  private heartbeat = () => {
    this.wssAlive = true;
  }

  private getRTTs = (): RTTs => {
    return {}
    // const rtts: RTTs = {}
    //
    // for (const conns of this.server.getConnections()) {
    //   rtts[conns.remotePeer.toString()] = conns.rtt ?? -1
    // }
    //
    // return rtts
  }

  private rttSetup = () => {
    const self = this

    setInterval(async () => {
      self.lastRTTs = self.getRTTs()
    }, this.rttUpdateIntervalMs)
  }

  private getStreams = (): Streams => {
    const connections = this.server.getConnections();

    return connections.reduce((accumulator, connection) => {
      // Ensure the connection has streams and a valid remotePeer
      if (connection.streams && connection.streams.length > 0 && connection.remotePeer) {
        const peerId = connection.remotePeer.toString();

        // Initialize the array for this peerId if it doesn't exist
        if (!accumulator[peerId]) {
          accumulator[peerId] = [];
        }

        // Map the streams to the desired format and append them to the peer's array
        const mappedStreams: Stream[] = connection.streams.map(stream => ({
          protocol: stream.protocol || '',
          direction: stream.direction
        }));

        accumulator[peerId].push(...mappedStreams);
      }

      return accumulator;
    }, {} as Streams); // Initialize accumulator as an empty object
  };
}

File: go-nodes/bootstrapper/internal/statusserver/statusserver.go
package statusserver

import (
	"encoding/json"
	"fmt"
	"net/http"
	"sync"
	"time"

	"github.com/dozyio/gossipsub-simulator/go-nodes/bootstrapper/internal/node"
	"github.com/gorilla/websocket"
	logging "github.com/ipfs/go-log/v2"
	"github.com/libp2p/go-libp2p/core/peer"
	"github.com/tidwall/gjson"
)

var upgrader = websocket.Upgrader{
	CheckOrigin: func(r *http.Request) bool { return true },
}

type Stream struct {
	Protocol  string `json:"protocol"`
	Direction string `json:"direction"`
}

type Streams map[string][]Stream

type PeerScores map[string]int

type RTTs map[string]int

type KadPeer struct {
	KadId    []byte  `json:"kadId"`
	PeerId   peer.ID `json:"peerId"`
	LastPing int64   `json:"lastPing"`
}

type RemotePeer struct {
	Multiaddrs []string `json:"multiaddrs"`
	Protocols  []string `json:"protocols"`
}

type RemotePeers map[string]*RemotePeer

type DHTProvideStatus string

const (
	DHTProvideError      DHTProvideStatus = "error"
	DHTProvideInProgress DHTProvideStatus = "inprogress"
	DHTProvideDone       DHTProvideStatus = "done"
)

type TopicsPeers map[string][]string

type Update struct {
	ContainerId           string           `json:"containerId"`
	PeerId                string           `json:"peerId"`
	Type                  string           `json:"type"`
	SubscribersList       TopicsPeers      `json:"subscribersList"`
	PubsubPeers           []string         `json:"pubsubPeers"`
	MeshPeersList         TopicsPeers      `json:"meshPeersList"`
	FanoutList            TopicsPeers      `json:"fanoutList"`
	Libp2pPeers           []string         `json:"libp2pPeers"`
	Connections           []string         `json:"connections"`
	RemotePeers           RemotePeers      `json:"remotePeers"`
	Protocols             []string         `json:"protocols"`
	Streams               Streams          `json:"streams"`
	Multiaddrs            []string         `json:"multiaddrs"`
	Topics                []string         `json:"topics"`
	DhtPeers              []string         `json:"dhtPeers"`
	DhtProvideStatus      DHTProvideStatus `json:"dhtProvideStatus"`
	DhtFindProviderResult []string         `json:"dhtFindProviderResult"`
	DhtFindPeerResult     []string         `json:"dhtFindPeerResult"`
	LastMessage           string           `json:"lastMessage"`
	MessageCount          int              `json:"messageCount"`
	PeerScores            PeerScores       `json:"peerScores"`
	RTTs                  RTTs             `json:"rtts"`
	ConnectTime           int64            `json:"connectTime"`
}

type StatusServer struct {
	Node        *node.Node
	Type        string
	PeerId      peer.ID
	Topics      []string
	ContainerId string
	PerfBytes   int

	clients   map[*websocket.Conn]bool
	clientsMu sync.RWMutex

	LastPeerId      string
	LastType        string
	LastLibp2pPeers []string
	LastConnections []string
	LastRemotePeers RemotePeers
	LastProtocols   []string
	LastStreams     Streams
	LastMultiaddrs  []string
	LastRTTs        RTTs

	LastTopics          []string
	LastSubscribersList TopicsPeers
	LastPubsubPeers     []string
	LastMeshPeersList   TopicsPeers
	LastFanoutList      TopicsPeers
	LastMessage         string
	MessageCount        int
	LastPeerScores      PeerScores

	LastDhtPeers              []string
	LastDhtProvideStatus      DHTProvideStatus
	LastDhtFindProviderResult []string
	LastDhtFindPeerResult     []string
	ConnectTime               int64
	LastConnectTime           int64

	WssAlive              bool
	Started               bool
	ScheduleUpdate        *time.Ticker
	UpdateCount           int
	UpdateIntervalMs      int
	UpdatesBeforeFullData int
	RttUpdateIntervalMs   int
	Wss                   *websocket.Conn
	WssMutex              sync.Mutex
	log                   logging.EventLogger
}

func New(n *node.Node, typ string, topics []string, perfBytes int, log logging.EventLogger) *StatusServer {
	ss := &StatusServer{
		Node:                  n,
		Type:                  typ,
		PeerId:                n.Host.ID(),
		Topics:                topics,
		PerfBytes:             perfBytes,
		WssAlive:              false,
		Started:               false,
		UpdateIntervalMs:      100,
		UpdatesBeforeFullData: 100,
		RttUpdateIntervalMs:   1000,
		log:                   log,
		clients:               make(map[*websocket.Conn]bool),
	}

	// Event handlers
	ss.serverEventHandlers()

	mux := http.NewServeMux()
	mux.HandleFunc("/", ss.wsHandler)

	handler := enableCors(mux)

	srv := &http.Server{
		Addr:              ":80",
		Handler:           handler,
		ReadTimeout:       5 * time.Second,
		WriteTimeout:      7 * time.Second,
		ReadHeaderTimeout: 3 * time.Second,
		IdleTimeout:       60 * time.Second,
	}

	ss.scheduleUpdates()

	go func() {
		log.Fatal(srv.ListenAndServe())
	}()

	return ss
}

func enableCors(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type")

		if r.Method == "OPTIONS" {
			w.Header().Set("Access-Control-Max-Age", "3600")
			return
		}

		next.ServeHTTP(w, r)
	})
}

func (ss *StatusServer) wsHandler(w http.ResponseWriter, r *http.Request) {
	ss.log.Debugf("new statusserver conn: %s", r.RemoteAddr)
	c, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		ss.log.Debugf("upgrade failed: %s", err)
		return
	}

	go func() {
		ss.clientsMu.Lock()
		{
			ss.clients[c] = true
		}
		ss.clientsMu.Unlock()

		defer func() {
			ss.clientsMu.Lock()
			{
				delete(ss.clients, c)
			}
			ss.clientsMu.Unlock()
			c.Close()
		}()

		// send full update on connection
		fullUpdate, err := ss.fullUpdate()
		if err != nil {
			ss.log.Warnf("error getting full update", err)
		} else {
			err = ss.sendUpdate(c, fullUpdate)
			if err != nil {
				ss.log.Warnf("error sending update, closing connection %s", err)
				return
			}
		}

		for {
			_, b, err := c.ReadMessage()
			if err != nil {
				ss.log.Debugf("disconnected: %s %s", c.RemoteAddr(), err)
				// Client disconnected
				return
			}

			fmt.Printf("Received message: %s\n", string(b))

			mType := gjson.GetBytes(b, "mType").String()
			switch mType {
			case "set-id":
				ss.log.Infof("setting container id %s", gjson.GetBytes(b, "message.id").String())
				ss.ContainerId = gjson.GetBytes(b, "message.id").String()

				fullUpdate, err := ss.fullUpdate()
				if err != nil {
					ss.log.Warnf("error getting full update", err)
					continue
				}

				err = ss.sendUpdate(c, fullUpdate)
				if err != nil {
					ss.log.Warnf("error sending update, closing connection %s", err)
					return
				}
			}
		}
	}()
}

func (ss *StatusServer) setupWSS() {
	// Configure the WebSocket server
}

func (ss *StatusServer) serverEventHandlers() {
	// Setup event listeners from libp2p server (connection:open, etc.)
}

func (ss *StatusServer) handleSelfPeerUpdate() {
	// Handle peer update
}

func (ss *StatusServer) handleConnectionEvent() {
	// Handle connection event
}

func (ss *StatusServer) handlePubsubMessageEvent() {
	// Handle pubsub message
}

func (ss *StatusServer) getRemotePeers() (*RemotePeers, error) {
	connsRaw := ss.Node.Host.Network().Conns()
	remotePeers := make(RemotePeers)

	for _, c := range connsRaw {
		rAddrs := ss.Node.Host.Peerstore().PeerInfo(c.RemotePeer()).Addrs

		rawProtocols, err := ss.Node.Host.Peerstore().GetProtocols(c.RemotePeer())
		if err != nil {
			ss.log.Errorf("error getting protocols for remote peer: %s", err)
		}

		rp := &RemotePeer{
			Multiaddrs: []string{},
			Protocols:  []string{},
		}

		for _, rAddr := range rAddrs {
			rp.Multiaddrs = append(rp.Multiaddrs, rAddr.String())
		}

		for _, protocol := range rawProtocols {
			rp.Protocols = append(rp.Protocols, string(protocol))
		}

		remotePeers[c.RemotePeer().String()] = rp
	}

	return &remotePeers, nil
}

func (ss *StatusServer) getPeerScores() PeerScores {
	// Go doesn't expose the peer scores
	return PeerScores{}
}

func (ss *StatusServer) getRTTs() RTTs {
	// Go doesn't track RTTs, return -1
	rtts := RTTs{}

	connsRaw := ss.Node.Host.Network().Conns()
	for _, c := range connsRaw {
		rtts[c.RemotePeer().String()] = -1
	}

	// Get round-trip times
	return rtts
}

func (ss *StatusServer) deltaUpdate() (*Update, error) {
	// Perform a delta update
	// TODO
	return &Update{}, nil
}

func (ss *StatusServer) fullUpdate() (*Update, error) {
	// Perform a full update
	mas := ss.Node.Host.Addrs()
	maStrings := make([]string, len(mas))
	for i, ma := range mas {
		maStrings[i] = ma.String()
	}

	protocolsRaw := ss.Node.Host.Mux().Protocols()
	protocols := make([]string, len(protocolsRaw))
	for i, p := range protocolsRaw {
		protocols[i] = string(p)
	}

	peersRaw := ss.Node.Host.Peerstore().Peers()
	peers := make([]string, len(peersRaw))
	for i, p := range peersRaw {
		peers[i] = p.String()
	}

	connsRaw := ss.Node.Host.Network().Conns()
	conns := make([]string, len(connsRaw))
	for i, c := range connsRaw {
		conns[i] = c.RemotePeer().String()
	}

	remotePeers, err := ss.getRemotePeers()
	if err != nil {
		ss.log.Errorf("error getting remote peers: %s", err)
		return nil, err
	}

	subscribersList := ss.getSubscribersList()
	meshPeersList := ss.getMeshPeersList() // empty
	rtts := ss.getRTTs()                   // not tracked by Go, return -1

	u := &Update{
		ContainerId:           ss.ContainerId,
		PeerId:                ss.Node.Host.ID().String(),
		Type:                  ss.Type,
		SubscribersList:       subscribersList,
		MeshPeersList:         meshPeersList,
		Libp2pPeers:           peers,
		Connections:           conns,
		RemotePeers:           *remotePeers,
		Protocols:             protocols,
		Streams:               Streams{},
		Multiaddrs:            maStrings,
		Topics:                ss.Topics,
		DhtPeers:              []string{},
		DhtProvideStatus:      "",
		DhtFindProviderResult: []string{},
		DhtFindPeerResult:     []string{},
		LastMessage:           "",
		MessageCount:          0,
		PeerScores:            PeerScores{},
		RTTs:                  rtts,
		ConnectTime:           0,
	}
	return u, nil
}

func (ss *StatusServer) sendUpdate(c *websocket.Conn, update *Update) error {
	data, err := json.Marshal(update)
	if err != nil {
		ss.log.Warnf("error sendUpdate marshaling update", err)
		return err
	}

	err = c.WriteMessage(websocket.TextMessage, data)
	if err != nil {
		return err
	}

	return nil
}

// Schedule periodic updates
func (ss *StatusServer) scheduleUpdates() {
	go func() {
		for {
			u, err := ss.fullUpdate()
			if err != nil {
				ss.log.Warnf("error getting full update", err)
				return
			}

			ss.clientsMu.Lock()
			for c := range ss.clients {
				err := ss.sendUpdate(c, u)
				if err != nil {
					fmt.Printf("Failed to write message to client %s: %v\n", c.RemoteAddr().String(), err)
					c.Close()
					delete(ss.clients, c)
				}
			}
			ss.clientsMu.Unlock()

			time.Sleep(5 * time.Second)
		}
	}()
}

func (ss *StatusServer) getStreams() Streams {
	// Get active streams from connections
	// TODO
	return Streams{}
}

func (ss *StatusServer) getMeshPeersList() TopicsPeers {
	// Go doesn't expose the mesh peers list
	return TopicsPeers{}
}

func (ss *StatusServer) getSubscribersList() TopicsPeers {
	// only handle 1 topic for now
	topicsPeers := make(TopicsPeers)

	peers := ss.Node.Topic.ListPeers()
	ss.log.Infof("subscribers topiclistpeers peers: len: %d, %+v", len(peers), peers)

	var peerIds []string
	for _, p := range peers {
		peerIds = append(peerIds, p.String())
	}

	topicsPeers[ss.Node.Topic.String()] = peerIds

	ss.log.Infof("subscribers peers: %+v", topicsPeers)

	return topicsPeers
}
